substrate/frame/asset-rewards/src/lib.rs

```diff
@@ -0,0 +1,713 @@
+			pool_id: PoolId,
+			new_expiry_block: BlockNumberFor<T>,
+		) -> DispatchResult {
+			let caller = T::PermissionedOrigin::ensure_origin(origin.clone())
+				.or_else(|_| ensure_signed(origin))?;
+
+			ensure!(
+				new_expiry_block > frame_system::Pallet::<T>::block_number(),
+				Error::<T>::ExpiryBlockMustBeInTheFuture
+			);
+
+			// Always start by updating the pool rewards.
+			let pool_info = Pools::<T>::get(pool_id).ok_or(Error::<T>::NonExistentPool)?;
+			let mut pool_info = Self::update_pool_rewards(pool_info)?;
+
+			ensure!(pool_info.admin == caller, BadOrigin);
+			pool_info.expiry_block = new_expiry_block;
+			Pools::<T>::insert(pool_id, pool_info);
+
+			Self::deposit_event(Event::PoolExpiryBlockModified { pool_id, new_expiry_block });
+
+			Ok(())
+		}
+
+		/// Convinience method to deposit reward tokens into a pool.
+		///
+		/// This method is not strictly necessary (tokens could be transferred directly to the
+		/// pool pot address), but is provided for convenience so manual derivation of the
+		/// account id is not required.
+		#[pallet::call_index(7)]
+		pub fn deposit_reward_tokens(
+			origin: OriginFor<T>,
+			pool_id: PoolId,
+			amount: T::Balance,
+		) -> DispatchResult {
+			let caller = ensure_signed(origin)?;
+			let pool_info = Pools::<T>::get(pool_id).ok_or(Error::<T>::NonExistentPool)?;
+			let pool_account_id = Self::pool_account_id(&pool_id)?;
+			T::Assets::transfer(
+				pool_info.reward_asset_id,
+				&caller,
+				&pool_account_id,
+				amount,
+				Preservation::Preserve,
+			)?;
+			Ok(())
+		}
+
+		/// Permissioned method to withdraw reward tokens from a pool.
+		#[pallet::call_index(8)]
+		pub fn withdraw_reward_tokens(
+			origin: OriginFor<T>,
+			pool_id: PoolId,
+			amount: T::Balance,
+			dest: T::AccountId,
+		) -> DispatchResult {
+			let caller = T::PermissionedOrigin::ensure_origin(origin.clone())
+				.or_else(|_| ensure_signed(origin))?;
+
+			let pool_info = Pools::<T>::get(pool_id).ok_or(Error::<T>::NonExistentPool)?;
+			ensure!(pool_info.admin == caller, BadOrigin);
+
+			T::Assets::transfer(
+				pool_info.reward_asset_id,
+				&Self::pool_account_id(&pool_id)?,
+				&dest,
+				amount,
+				Preservation::Preserve,
+			)?;
+
+			Ok(())
+		}
+	}
+
+	impl<T: Config> Pallet<T> {
+		/// Derive a pool account ID from the pallet's ID.
+		pub fn pool_account_id(id: &PoolId) -> Result<T::AccountId, DispatchError> {
+			if Pools::<T>::contains_key(id) {
+				Ok(T::PalletId::get().into_sub_account_truncating(id))
+			} else {
+				Err(Error::<T>::NonExistentPool.into())
+			}
+		}
+
+		/// Computes update pool and staker reward state.
+		///
+		/// Should be called prior to any operation involving a staker.
+		///
+		/// Returns the updated pool and staker info.
+		///
+		/// NOTE: does not modify any storage, that is the responsibility of the caller.
+		pub fn update_pool_and_staker_rewards(
+			pool_info: PoolInfoFor<T>,
+			mut staker_info: PoolStakerInfo<T::Balance>,
```

Comment:
it would be nicer if we do not mutate staker and pool infos in these function, and do mutate only in a single context (call function for example). With the current approach the state of these objects in not clear if you do not follow all the path. These function could return calculated values to the caller.

Reply:
I was thinking that too, but feel like it's ok because the method completely consumes `staker_info`, so it can't be access after being passed in here.

by passing it non-mut, I think we would just be requiring a useless clone. but let me know if I'm wrong.

Reply:
Cloning data is not significant concern here, even when you move the data you do, under the hood it will be most probably copied.
This is why maintenance is a bigger factor here. And pure mathematical functions easier to maintain.

Reply:
cool, yeah I initially had them as pure functions and I agree from a code maintainability and understanding point is better, will update :)

Reply:
@muharem I've simplified the logic a little, lmk if looks better